File: WinForms/CS/PivotGridMainDemo/Modules/CalculatedFields.cs
using System;
using System.Collections;
using System.Linq;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.Data;
using DevExpress.Utils.Menu;
using DevExpress.XtraEditors;

namespace DevExpress.XtraPivotGrid.Demos.Modules {
    public partial class CalculatedFields : DevExpress.XtraPivotGrid.Demos.TutorialControl {
        PivotGridField hoverField = null;

        public CalculatedFields() {
            // This call is required by the Windows Form Designer.
            CreateWaitDialog();
            InitializeComponent();

            pivotGridControl.BeginUpdate();

            fieldYear.FilterValues.SetValues(new object[] { 2015 }, PivotFilterType.Included, false);

            checkExpression.Checked = true;

            SetUpYearEndBonus();
            SetUpQuantityBonus();

            pivotGridControl.EndUpdate();
        }
        public override PivotGridControl ViewOptionsControl {
            get { return pivotGridControl; }
        }
        public override PivotGridControl ExportControl {
            get { return pivotGridControl; }
        }
        void CalculatedFields_Load(object sender, System.EventArgs e) {
            pivotGridControl.DataSource = GetData();
            pivotGridControl.HeaderImages = HeadersImageList;
            fieldSalesPerson.ImageIndex = 0;
        }
        List<SalesDataRow> GetData() {
            Dictionary<int, DataRowView> orders = GetNWindData("Orders").Cast<DataRowView>().ToDictionary(s => (int)s["OrderID"], s => s);
            Dictionary<int, DataRowView> employees = GetNWindData("Employees").Cast<DataRowView>().ToDictionary(s => (int)s["EmployeeID"], s => s);
            return GetNWindData("Order Details").Cast<DataRowView>().Select(s => CreateSalesDataRow(s, orders, employees)).ToList();
        }

        static SalesDataRow CreateSalesDataRow(DataRowView order, Dictionary<int, DataRowView> orders, Dictionary<int, DataRowView> employees) {
            int orderId = (int)order["OrderID"];
            int employeeID = (int)orders[orderId]["EmployeeID"];
            DataRowView employee = employees[employeeID];

            return new SalesDataRow {
                OrderID = (int)order["OrderID"],
                UnitPrice = (decimal)order["UnitPrice"],
                Quantity = (short)order["Quantity"],

                OrderDate = (DateTime)orders[orderId]["OrderDate"],

                LastName = (string)employee["LastName"],
                FirstName = (string)employee["FirstName"],
                EmployeeID = (int)employee["EmployeeID"],
            };
        }

        public class SalesDataRow {
            public int EmployeeID { get; set; }
            public int OrderID { get; set; }
            public decimal UnitPrice { get; set; }
            public short Quantity { get; set; }
            public DateTime OrderDate { get; set; }
            public string FirstName { get; set; }
            public string LastName { get; set; }
        }

        //<checkDataSourceValue>
        void checkDataSourceValueOnCheckedChanged(object sender, EventArgs e) {
            if(!checkDataSourceValue.Checked)
                return;
            fieldSalesPerson.DataBinding = new DataSourceColumnBinding("EmployeeID");
            pivotGridControl.RefreshData();
        }
        //</checkDataSourceValue>
        //<checkDisplayText>
        void checkDisplayTextOnCheckedChanged(object sender, EventArgs e) {
            if(!checkDisplayText.Checked)
                return;
            fieldSalesPerson.DataBinding = new DataSourceColumnBinding("EmployeeID");
            pivotGridControl.RefreshData();
        }
        void pivotGridControlOnFieldValueDisplayText(object sender, PivotFieldDisplayTextEventArgs e) {
            if(!checkDisplayText.Checked)
                return;
            if(e.Field == fieldSalesPerson) {
                PivotDrillDownDataSource ds = e.CreateDrillDownDataSource();
                e.DisplayText = string.Format("{0} {1} ({2})", ds[0]["FirstName"], ds[0]["LastName"], e.Value.ToString());
            }
        }
        //</checkDisplayText>
        //<checkExpression>
        void checkExpressionOnCheckedChanged(object sender, EventArgs e) {
            if(!checkExpression.Checked)
                return;
            fieldSalesPerson.DataBinding = new ExpressionDataBinding(string.Format("Concat([{0}], ' ', [{1}], ' (', [{2}], ')')", fieldFirstName.Name, fieldLastName.Name, fieldEmployeeID.Name));
            fieldSalesPerson.Options.ShowExpressionEditorMenu = true;
            fieldSalesPerson.FieldName = "4";
            pivotGridControl.RefreshData();
        }
        //</checkExpression>
        //<checkYearEndBonus>
        void SetUpYearEndBonus() {
            fieldYearEndBonus.Options.ShowExpressionEditorMenu = true;
            checkYearEndBonus.Checked = true;
        }
        void checkYearEndBonusOnCheckedChanged(object sender, EventArgs e) {
            fieldYearEndBonus.Visible = checkYearEndBonus.Checked;
            if(fieldYearEndBonus.Visible)
                fieldYearEndBonus.DataBinding = new ExpressionDataBinding(string.Format("Iif( Max([{0}]) == 4 And Min([{0}]) == 4, Iif(Sum([{1}]) > 40000, 600, 400), 0)", fieldQuarter.Name, fieldExtendedPrice.Name));
            else
                fieldYearEndBonus.DataBinding = new ExpressionDataBinding("0");
        }
        //</checkYearEndBonus>
        //<checkQuantityBonus>
        void SetUpQuantityBonus() {
            fieldQuantityBonus.Options.ShowExpressionEditorMenu = true;
            checkQuantityBonus.Checked = true;
        }
        void checkQuantityBonusOnCheckedChanged(object sender, EventArgs e) {
            fieldQuantityBonus.Visible = checkQuantityBonus.Checked;
            if(fieldQuantityBonus.Visible)
                fieldQuantityBonus.DataBinding = new ExpressionDataBinding(string.Format("Iif([{0}] > 100, [{1}] * 0.1, 0)", fieldQuantity.Name, fieldExtendedPrice.Name));
            else
                fieldQuantityBonus.DataBinding = new ExpressionDataBinding("0");
        }
        //</checkQuantityBonus>
        //<pivotGridControl>
        void pivotGridControlOnCellDoubleClick(object sender, PivotCellEventArgs e) {
            try {
                using(DrillDownForm form = new DrillDownForm(e.CreateDrillDownDataSource())) {
                    form.ShowDialog();
                }
            }
            catch(Exception ex) {
                XtraMessageBox.Show(ex.Message);
            }
        }
        //</pivotGridControl>
        //<btnAddBonus>
        void buttonAddNewBonusOnClick(object sender, EventArgs e) {
            using(AddNewBonus addNewBonus = new AddNewBonus()) {
                addNewBonus.PivotGrid = pivotGridControl;
                addNewBonus.ShowDialog();
            }
        }
        //</btnAddBonus>

        bool NeedHighlightField(PivotGridField field) {
            if(field != null && field.DataBinding is ExpressionDataBinding && field.Options.ShowExpressionEditorMenu)
                return true;
            return false;
        }
        void pivotGridControlOnMouseMove(object sender, MouseEventArgs e) {
            PivotGridHitInfo info = pivotGridControl.CalcHitInfo(new Point(e.X, e.Y));
            if(info.ValueInfo != null && NeedHighlightField(info.ValueInfo.Field) && info.ValueInfo.Field.Area == PivotArea.DataArea) {
                hoverField = info.ValueInfo.Field;
                pivotGridControl.Cursor = Cursors.Hand;
            }
            else if(info.HitTest == PivotGridHitTest.HeadersArea && NeedHighlightField(info.HeaderField)) {
                hoverField = info.HeaderField;
                pivotGridControl.Cursor = Cursors.Hand;
            }
            else {
                hoverField = null;
                pivotGridControl.Cursor = Cursors.Default;
            }
            pivotGridControl.Invalidate();
        }
        void pivotGridControlOnCustomDrawFieldValue(object sender, PivotCustomDrawFieldValueEventArgs e) {
            if(NeedHighlightField(e.Field) && e.Field.Area == PivotArea.DataArea) {
                if(e.Field == hoverField)
                    e.Appearance.FontStyleDelta = FontStyle.Underline | FontStyle.Bold;
                else
                    e.Appearance.FontStyleDelta = FontStyle.Bold;
            }
            else e.DefaultDraw();
        }
        void pivotGridControlOnCustomDrawFieldHeader(object sender, PivotCustomDrawFieldHeaderEventArgs e) {
            if(NeedHighlightField(e.Field)) {
                if(e.Field == hoverField)
                    e.Appearance.FontStyleDelta = FontStyle.Underline | FontStyle.Bold;
                else
                    e.Appearance.FontStyleDelta = FontStyle.Bold;
            }
            else e.DefaultDraw();
        }
        void pivotGridControlOnMouseClick(object sender, MouseEventArgs e) {
            if(e.Button != MouseButtons.Left)
                return;
            PivotGridHitInfo info = pivotGridControl.CalcHitInfo(new Point(e.X, e.Y));
            if(info.ValueInfo != null && NeedHighlightField(info.ValueInfo.Field) && info.ValueInfo.Field.Area == PivotArea.DataArea)
                pivotGridControl.ShowExpressionEditor(info.ValueInfo.Field);
            else if(NeedHighlightField(info.HeaderField))
                pivotGridControl.ShowExpressionEditor(info.HeaderField);
        }
        void pivotGridControlOnPopupMenuShowing(object sender, PopupMenuShowingEventArgs args) {
            if(args.HitInfo.ValueInfo != null) {
                PivotGridField removedField = args.HitInfo.ValueInfo.Field;
                if(removedField != null && Equals(removedField.Tag, "removable")) {
                    DXMenuItem item = new DXMenuItem("Remove this bonus");
                    item.Click += (s, e) => {
                        pivotGridControl.Fields.Remove(removedField);
                    };
                    item.BeginGroup = true;
                    args.Menu.Items.Add(item);
                }
            }
        }
    }
}

File: WinForms/VB/PivotGridMainDemo/Modules/CalculatedFields.vb
Imports System
Imports System.Linq
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraEditors

Namespace DevExpress.XtraPivotGrid.Demos.Modules

    Public Partial Class CalculatedFields
        Inherits TutorialControl

        Private hoverField As PivotGridField = Nothing

        Public Sub New()
            ' This call is required by the Windows Form Designer.
            CreateWaitDialog()
            InitializeComponent()
            pivotGridControl.BeginUpdate()
            fieldYear.FilterValues.SetValues(New Object() {2015}, PivotFilterType.Included, False)
            checkExpression.Checked = True
            SetUpYearEndBonus()
            SetUpQuantityBonus()
            pivotGridControl.EndUpdate()
        End Sub

        Public Overrides ReadOnly Property ViewOptionsControl As PivotGridControl
            Get
                Return pivotGridControl
            End Get
        End Property

        Public Overrides ReadOnly Property ExportControl As PivotGridControl
            Get
                Return pivotGridControl
            End Get
        End Property

        Private Sub CalculatedFields_Load(ByVal sender As Object, ByVal e As EventArgs)
            pivotGridControl.DataSource = GetData()
            pivotGridControl.HeaderImages = HeadersImageList
            fieldSalesPerson.ImageIndex = 0
        End Sub

        Private Function GetData() As List(Of SalesDataRow)
            Dim orders As Dictionary(Of Integer, DataRowView) = GetNWindData("Orders").Cast(Of DataRowView)().ToDictionary(Function(s) CInt(s("OrderID")), Function(s) s)
            Dim employees As Dictionary(Of Integer, DataRowView) = GetNWindData("Employees").Cast(Of DataRowView)().ToDictionary(Function(s) CInt(s("EmployeeID")), Function(s) s)
            Return GetNWindData("Order Details").Cast(Of DataRowView)().[Select](Function(s) CreateSalesDataRow(s, orders, employees)).ToList()
        End Function

        Private Shared Function CreateSalesDataRow(ByVal order As DataRowView, ByVal orders As Dictionary(Of Integer, DataRowView), ByVal employees As Dictionary(Of Integer, DataRowView)) As SalesDataRow
            Dim orderId As Integer = CInt(order("OrderID"))
            Dim employeeID As Integer = CInt(orders(orderId)("EmployeeID"))
            Dim employee As DataRowView = employees(employeeID)
            Return New SalesDataRow With {.OrderID = CInt(order("OrderID")), .UnitPrice = CDec(order("UnitPrice")), .Quantity = CShort(order("Quantity")), .OrderDate = CDate(orders(orderId)("OrderDate")), .LastName = CStr(employee("LastName")), .FirstName = CStr(employee("FirstName")), .EmployeeID = CInt(employee("EmployeeID"))}
        End Function

        Public Class SalesDataRow

            Public Property EmployeeID As Integer

            Public Property OrderID As Integer

            Public Property UnitPrice As Decimal

            Public Property Quantity As Short

            Public Property OrderDate As Date

            Public Property FirstName As String

            Public Property LastName As String
        End Class

        '<checkDataSourceValue>
        Private Sub checkDataSourceValueOnCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If Not checkDataSourceValue.Checked Then Return
            fieldSalesPerson.DataBinding = New DataSourceColumnBinding("EmployeeID")
            pivotGridControl.RefreshData()
        End Sub

        '</checkDataSourceValue>
        '<checkDisplayText>
        Private Sub checkDisplayTextOnCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If Not checkDisplayText.Checked Then Return
            fieldSalesPerson.DataBinding = New DataSourceColumnBinding("EmployeeID")
            pivotGridControl.RefreshData()
        End Sub

        Private Sub pivotGridControlOnFieldValueDisplayText(ByVal sender As Object, ByVal e As PivotFieldDisplayTextEventArgs)
            If Not checkDisplayText.Checked Then Return
            If e.Field Is fieldSalesPerson Then
                Dim ds As PivotDrillDownDataSource = e.CreateDrillDownDataSource()
                e.DisplayText = String.Format("{0} {1} ({2})", ds(0)("FirstName"), ds(0)("LastName"), e.Value.ToString())
            End If
        End Sub

        '</checkDisplayText>
        '<checkExpression>
        Private Sub checkExpressionOnCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If Not checkExpression.Checked Then Return
            fieldSalesPerson.DataBinding = New ExpressionDataBinding(String.Format("Concat([{0}], ' ', [{1}], ' (', [{2}], ')')", fieldFirstName.Name, fieldLastName.Name, fieldEmployeeID.Name))
            fieldSalesPerson.Options.ShowExpressionEditorMenu = True
            fieldSalesPerson.FieldName = "4"
            pivotGridControl.RefreshData()
        End Sub

        '</checkExpression>
        '<checkYearEndBonus>
        Private Sub SetUpYearEndBonus()
            fieldYearEndBonus.Options.ShowExpressionEditorMenu = True
            checkYearEndBonus.Checked = True
        End Sub

        Private Sub checkYearEndBonusOnCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            fieldYearEndBonus.Visible = checkYearEndBonus.Checked
            If fieldYearEndBonus.Visible Then
                fieldYearEndBonus.DataBinding = New ExpressionDataBinding(String.Format("Iif( Max([{0}]) == 4 And Min([{0}]) == 4, Iif(Sum([{1}]) > 40000, 600, 400), 0)", fieldQuarter.Name, fieldExtendedPrice.Name))
            Else
                fieldYearEndBonus.DataBinding = New ExpressionDataBinding("0")
            End If
        End Sub

        '</checkYearEndBonus>
        '<checkQuantityBonus>
        Private Sub SetUpQuantityBonus()
            fieldQuantityBonus.Options.ShowExpressionEditorMenu = True
            checkQuantityBonus.Checked = True
        End Sub

        Private Sub checkQuantityBonusOnCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            fieldQuantityBonus.Visible = checkQuantityBonus.Checked
            If fieldQuantityBonus.Visible Then
                fieldQuantityBonus.DataBinding = New ExpressionDataBinding(String.Format("Iif([{0}] > 100, [{1}] * 0.1, 0)", fieldQuantity.Name, fieldExtendedPrice.Name))
            Else
                fieldQuantityBonus.DataBinding = New ExpressionDataBinding("0")
            End If
        End Sub

        '</checkQuantityBonus>
        '<pivotGridControl>
        Private Sub pivotGridControlOnCellDoubleClick(ByVal sender As Object, ByVal e As PivotCellEventArgs)
            Try
                Using form As DrillDownForm = New DrillDownForm(e.CreateDrillDownDataSource())
                    form.ShowDialog()
                End Using
            Catch ex As Exception
                XtraMessageBox.Show(ex.Message)
            End Try
        End Sub

        '</pivotGridControl>
        '<btnAddBonus>
        Private Sub buttonAddNewBonusOnClick(ByVal sender As Object, ByVal e As EventArgs)
            Using addNewBonus As AddNewBonus = New AddNewBonus()
                addNewBonus.PivotGrid = pivotGridControl
                addNewBonus.ShowDialog()
            End Using
        End Sub

        '</btnAddBonus>
        Private Function NeedHighlightField(ByVal field As PivotGridField) As Boolean
            If field IsNot Nothing AndAlso TypeOf field.DataBinding Is ExpressionDataBinding AndAlso field.Options.ShowExpressionEditorMenu Then Return True
            Return False
        End Function

        Private Sub pivotGridControlOnMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim info As PivotGridHitInfo = pivotGridControl.CalcHitInfo(New Point(e.X, e.Y))
            If info.ValueInfo IsNot Nothing AndAlso NeedHighlightField(info.ValueInfo.Field) AndAlso info.ValueInfo.Field.Area = PivotArea.DataArea Then
                hoverField = info.ValueInfo.Field
                pivotGridControl.Cursor = Cursors.Hand
            ElseIf info.HitTest = PivotGridHitTest.HeadersArea AndAlso NeedHighlightField(info.HeaderField) Then
                hoverField = info.HeaderField
                pivotGridControl.Cursor = Cursors.Hand
            Else
                hoverField = Nothing
                pivotGridControl.Cursor = Cursors.Default
            End If

            pivotGridControl.Invalidate()
        End Sub

        Private Sub pivotGridControlOnCustomDrawFieldValue(ByVal sender As Object, ByVal e As PivotCustomDrawFieldValueEventArgs)
            If NeedHighlightField(e.Field) AndAlso e.Field.Area = PivotArea.DataArea Then
                If e.Field Is hoverField Then
                    e.Appearance.FontStyleDelta = FontStyle.Underline Or FontStyle.Bold
                Else
                    e.Appearance.FontStyleDelta = FontStyle.Bold
                End If
            Else
                e.DefaultDraw()
            End If
        End Sub

        Private Sub pivotGridControlOnCustomDrawFieldHeader(ByVal sender As Object, ByVal e As PivotCustomDrawFieldHeaderEventArgs)
            If NeedHighlightField(e.Field) Then
                If e.Field Is hoverField Then
                    e.Appearance.FontStyleDelta = FontStyle.Underline Or FontStyle.Bold
                Else
                    e.Appearance.FontStyleDelta = FontStyle.Bold
                End If
            Else
                e.DefaultDraw()
            End If
        End Sub

        Private Sub pivotGridControlOnMouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            If e.Button <> MouseButtons.Left Then Return
            Dim info As PivotGridHitInfo = pivotGridControl.CalcHitInfo(New Point(e.X, e.Y))
            If info.ValueInfo IsNot Nothing AndAlso NeedHighlightField(info.ValueInfo.Field) AndAlso info.ValueInfo.Field.Area = PivotArea.DataArea Then
                pivotGridControl.ShowExpressionEditor(info.ValueInfo.Field)
            ElseIf NeedHighlightField(info.HeaderField) Then
                pivotGridControl.ShowExpressionEditor(info.HeaderField)
            End If
        End Sub

        Private Sub pivotGridControlOnPopupMenuShowing(ByVal sender As Object, ByVal args As PopupMenuShowingEventArgs)
            If args.HitInfo.ValueInfo IsNot Nothing Then
                Dim removedField As PivotGridField = args.HitInfo.ValueInfo.Field
                If removedField IsNot Nothing AndAlso Equals(removedField.Tag, "removable") Then
                    Dim item As DXMenuItem = New DXMenuItem("Remove this bonus")
                    AddHandler item.Click, Sub(s, e) pivotGridControl.Fields.Remove(removedField)
                    item.BeginGroup = True
                    args.Menu.Items.Add(item)
                End If
            End If
        End Sub
    End Class
End Namespace
